# <center>  Working With Dataframes </center>
<div>
<img src="https://pandas.pydata.org/static/img/pandas.svg" width="600"/>
</div>

In the last lecture, we were introduced to the `pandas` library, which gives us the ability to work with two new data types: dataframes and series. In this lecture, we will learn more about how to use and work with dataframes.

We're going to talk about some ways to handle and work with dataframes, which includes basic selection and deletion of columns, renaming columns, etc. Then, we'll go a bit into subsetting data using various selection methods. Finally, we talk about some methods for dataframes that will be useful for data organization, manipulation and analysis.

## Handling Dataframes

To learn how to use dataframes to their fullest capacity, we first need to know how to handle them. First, let's import pandas and numpy.

In [1]:
import pandas as pd
import numpy as np

In [5]:
sales = pd.read_csv('../datasets/sales.csv')
sales.head(3)

Row ID        Order ID Order Date   Ship Date  Ship Mode Customer ID  \
0       1  CA-2017-152156    8/11/17    11/11/17          2    CG-12520   
1       2  CA-2017-152156    8/11/17    11/11/17          2    CG-12520   
2       3  CA-2017-138688    12/6/17  16/06/2017          2    DV-13045   

     Customer Name    Segment        Country         City       State  \
0      Claire Gute   Consumer  United States    Henderson    Kentucky   
1      Claire Gute   Consumer  United States    Henderson    Kentucky   
2  Darrin Van Huff  Corporate  United States  Los Angeles  California   

   Postal Code Region       Product ID         Category Sub-Category  \
0      42420.0  South  FUR-BO-10001798        Furniture    Bookcases   
1      42420.0  South  FUR-CH-10000454        Furniture       Chairs   
2      90036.0   West  OFF-LA-10000240  Office Supplies       Labels   

                                        Product Name   Sales  
0                  Bush Somerset Collection Bookcase  261.96  
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.94  
2  Self-Adhesive Address Labels for Typewriters b...   14.62

This is a dataset of sales from a retail store over 4 years. At a quick glance, we can see the type of data collected. We have some IDs in the form of alphanumerical data, some dates, names, locations, etc.

In getting acquainted with our data, we may want to quickly know how many entries there are and how many attributes are measured per entry:

In [6]:
print(sales.shape)

(9804, 18)


Recall that dataframes are made of concatanated series in the form of columns. To extract one column from a dataframe, we call for the dataframe and then close the name of the column (as a string) in single square brackets or double square brackets:

In [7]:
sales['Segment']

0        Consumer
1        Consumer
2       Corporate
3        Consumer
4        Consumer
          ...    
9799    Corporate
9800    Corporate
9801    Corporate
9802    Corporate
9803    Corporate
Name: Segment, Length: 9804, dtype: object

In [8]:
sales[['Segment']]

Segment
0      Consumer
1      Consumer
2     Corporate
3      Consumer
4      Consumer
...         ...
9799  Corporate
9800  Corporate
9801  Corporate
9802  Corporate
9803  Corporate

[9804 rows x 1 columns]

**What do notice about the difference in extracting a column with single square brackets versus double square brackets?**

Notice that using single square brackets returns a series, while the use of double square brackets return a one-column dataframe. This is important, as it can dictate how the output can be used, as well as what functions and methods can be used on it.

For example, if we wanted to get the first item of the `Segment` column, we could use `sales['Segment'][0]` with a series, but not with a dataframe:

In [9]:
sales['Segment'][0]

'Consumer'

In [10]:
sales[['Segment']][0]

KeyError: 0

However, if you wanted to select multiple columns at once, that can be done with a dataframe, but not a series:

In [11]:
sales[['Segment', 'Ship Mode']]

Segment  Ship Mode
0      Consumer          2
1      Consumer          2
2     Corporate          2
3      Consumer          3
4      Consumer          3
...         ...        ...
9799  Corporate          3
9800  Corporate          3
9801  Corporate          3
9802  Corporate          3
9803  Corporate          3

[9804 rows x 2 columns]

In [12]:
sales['Segment', 'Ship Mode']

KeyError: ('Segment', 'Ship Mode')

We can also extract a dataframe column by using a `dataframe.column_name` notation. This is best to use when there are no spaces in the column name. A series will be returned by using this notation, similar to `dataframe['column_name']`:

In [13]:
sales.Segment

0        Consumer
1        Consumer
2       Corporate
3        Consumer
4        Consumer
          ...    
9799    Corporate
9800    Corporate
9801    Corporate
9802    Corporate
9803    Corporate
Name: Segment, Length: 9804, dtype: object

It is also useful to rename columns in a dataframe using the `.rename()` method. Calling this method on a dataframe, you can easily rename a column by passing a dictionary with the keys being the current column name and the values being the desired column name:

In [14]:
sales = sales.rename(columns={'Postal Code':'Zip Code', 'Ship Mode':'Shipping Method'})
sales.head(3)

Row ID        Order ID Order Date   Ship Date  Shipping Method Customer ID  \
0       1  CA-2017-152156    8/11/17    11/11/17                2    CG-12520   
1       2  CA-2017-152156    8/11/17    11/11/17                2    CG-12520   
2       3  CA-2017-138688    12/6/17  16/06/2017                2    DV-13045   

     Customer Name    Segment        Country         City       State  \
0      Claire Gute   Consumer  United States    Henderson    Kentucky   
1      Claire Gute   Consumer  United States    Henderson    Kentucky   
2  Darrin Van Huff  Corporate  United States  Los Angeles  California   

   Zip Code Region       Product ID         Category Sub-Category  \
0   42420.0  South  FUR-BO-10001798        Furniture    Bookcases   
1   42420.0  South  FUR-CH-10000454        Furniture       Chairs   
2   90036.0   West  OFF-LA-10000240  Office Supplies       Labels   

                                        Product Name   Sales  
0                  Bush Somerset Collection Bookcase  261.96  
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.94  
2  Self-Adhesive Address Labels for Typewriters b...   14.62

New columns can be added to dataframes as well. If you wish to add a column based on other columns that already exist in the dataframe, you can perform operations on the columns and assign them to the new desired column. Here, we make a new column called `Full Location` that combines data from the `City` and `State` columns, separated by a comma:

In [15]:
sales['Full Location'] = sales['City'] + ', ' + sales['State']
sales.head(3)

Row ID        Order ID Order Date   Ship Date  Shipping Method Customer ID  \
0       1  CA-2017-152156    8/11/17    11/11/17                2    CG-12520   
1       2  CA-2017-152156    8/11/17    11/11/17                2    CG-12520   
2       3  CA-2017-138688    12/6/17  16/06/2017                2    DV-13045   

     Customer Name    Segment        Country         City       State  \
0      Claire Gute   Consumer  United States    Henderson    Kentucky   
1      Claire Gute   Consumer  United States    Henderson    Kentucky   
2  Darrin Van Huff  Corporate  United States  Los Angeles  California   

   Zip Code Region       Product ID         Category Sub-Category  \
0   42420.0  South  FUR-BO-10001798        Furniture    Bookcases   
1   42420.0  South  FUR-CH-10000454        Furniture       Chairs   
2   90036.0   West  OFF-LA-10000240  Office Supplies       Labels   

                                        Product Name   Sales  \
0                  Bush Somerset Collection Bookcase  261.96   
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.94   
2  Self-Adhesive Address Labels for Typewriters b...   14.62   

             Full Location  
0      Henderson, Kentucky  
1      Henderson, Kentucky  
2  Los Angeles, California

If desired, you can also create an empty column:

In [16]:
sales['Empty'] = ""
sales.head(3)

Row ID        Order ID Order Date   Ship Date  Shipping Method Customer ID  \
0       1  CA-2017-152156    8/11/17    11/11/17                2    CG-12520   
1       2  CA-2017-152156    8/11/17    11/11/17                2    CG-12520   
2       3  CA-2017-138688    12/6/17  16/06/2017                2    DV-13045   

     Customer Name    Segment        Country         City       State  \
0      Claire Gute   Consumer  United States    Henderson    Kentucky   
1      Claire Gute   Consumer  United States    Henderson    Kentucky   
2  Darrin Van Huff  Corporate  United States  Los Angeles  California   

   Zip Code Region       Product ID         Category Sub-Category  \
0   42420.0  South  FUR-BO-10001798        Furniture    Bookcases   
1   42420.0  South  FUR-CH-10000454        Furniture       Chairs   
2   90036.0   West  OFF-LA-10000240  Office Supplies       Labels   

                                        Product Name   Sales  \
0                  Bush Somerset Collection Bookcase  261.96   
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.94   
2  Self-Adhesive Address Labels for Typewriters b...   14.62   

             Full Location Empty  
0      Henderson, Kentucky        
1      Henderson, Kentucky        
2  Los Angeles, California

There may be times where you want to delete columns. This can be accomplished using the `.drop()` method. When you call this method on a dataframe, you can pass a single column name or a list of column names to the `columns` parameter. The `inplace=True` parameter updates the dataframe.

In [17]:
sales.drop(columns=['Country', 'Empty'], inplace=True)
sales.head(3)

Row ID        Order ID Order Date   Ship Date  Shipping Method Customer ID  \
0       1  CA-2017-152156    8/11/17    11/11/17                2    CG-12520   
1       2  CA-2017-152156    8/11/17    11/11/17                2    CG-12520   
2       3  CA-2017-138688    12/6/17  16/06/2017                2    DV-13045   

     Customer Name    Segment         City       State  Zip Code Region  \
0      Claire Gute   Consumer    Henderson    Kentucky   42420.0  South   
1      Claire Gute   Consumer    Henderson    Kentucky   42420.0  South   
2  Darrin Van Huff  Corporate  Los Angeles  California   90036.0   West   

        Product ID         Category Sub-Category  \
0  FUR-BO-10001798        Furniture    Bookcases   
1  FUR-CH-10000454        Furniture       Chairs   
2  OFF-LA-10000240  Office Supplies       Labels   

                                        Product Name   Sales  \
0                  Bush Somerset Collection Bookcase  261.96   
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.94   
2  Self-Adhesive Address Labels for Typewriters b...   14.62   

             Full Location  
0      Henderson, Kentucky  
1      Henderson, Kentucky  
2  Los Angeles, California

Notice that we utilized two ways of updating dataframes using these methods. One way was to use the method on the dataframe and reassign the output of that to the name of the dataframe (as we did when using the `.rename()` method).The other way was by specifying the `inplace` parameter as `True` (as we did with the `.drop()` method). 

Both of these ways are fine, but be mindful that not all methods and functions have an `inplace` parameter. Therefore, it's always a good idea to refer to the documentation of a function or method to understand how it can be used.

## Subsetting data in dataframes

Being able to parse and subset data intentionally is important for downstream data analysis. When consecutive rows of data need to be extracted, dataframes can be sliced similar to lists and arrays. To extract the first 50 rows of `sales` we could slice the dataframe as follows:

In [18]:
top50 = sales[0:50]
print('Number of rows in top50:', len(top50))
top50.head(3)

Number of rows in top50: 50


Row ID        Order ID Order Date   Ship Date  Shipping Method Customer ID  \
0       1  CA-2017-152156    8/11/17    11/11/17                2    CG-12520   
1       2  CA-2017-152156    8/11/17    11/11/17                2    CG-12520   
2       3  CA-2017-138688    12/6/17  16/06/2017                2    DV-13045   

     Customer Name    Segment         City       State  Zip Code Region  \
0      Claire Gute   Consumer    Henderson    Kentucky   42420.0  South   
1      Claire Gute   Consumer    Henderson    Kentucky   42420.0  South   
2  Darrin Van Huff  Corporate  Los Angeles  California   90036.0   West   

        Product ID         Category Sub-Category  \
0  FUR-BO-10001798        Furniture    Bookcases   
1  FUR-CH-10000454        Furniture       Chairs   
2  OFF-LA-10000240  Office Supplies       Labels   

                                        Product Name   Sales  \
0                  Bush Somerset Collection Bookcase  261.96   
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.94   
2  Self-Adhesive Address Labels for Typewriters b...   14.62   

             Full Location  
0      Henderson, Kentucky  
1      Henderson, Kentucky  
2  Los Angeles, California

Likewise we can extract intervals of rows using double colons:

In [19]:
even_rows = sales[0::2]
print('Selecting every other column gives us a dataframe of', len(even_rows), 'rows.')

even_rows.head(3)

Selecting every other column gives us a dataframe of 4902 rows.


Row ID        Order ID Order Date   Ship Date  Shipping Method Customer ID  \
0       1  CA-2017-152156    8/11/17    11/11/17                2    CG-12520   
2       3  CA-2017-138688    12/6/17  16/06/2017                2    DV-13045   
4       5  US-2016-108966   11/10/16  18/10/2016                3    SO-20335   

     Customer Name    Segment             City       State  Zip Code Region  \
0      Claire Gute   Consumer        Henderson    Kentucky   42420.0  South   
2  Darrin Van Huff  Corporate      Los Angeles  California   90036.0   West   
4   Sean O'Donnell   Consumer  Fort Lauderdale     Florida   33311.0  South   

        Product ID         Category Sub-Category  \
0  FUR-BO-10001798        Furniture    Bookcases   
2  OFF-LA-10000240  Office Supplies       Labels   
4  OFF-ST-10000760  Office Supplies      Storage   

                                        Product Name    Sales  \
0                  Bush Somerset Collection Bookcase  261.960   
2  Self-Adhesive Address Labels for Typewriters b...   14.620   
4                     Eldon Fold 'N Roll Cart System   22.368   

              Full Location  
0       Henderson, Kentucky  
2   Los Angeles, California  
4  Fort Lauderdale, Florida

To exact all rows and a specific column, we can use the `.iloc[]` method. The `.iloc[]` method uses integer-location based indexing for selection of rows and columns by position. To select all rows and the `Ship Date` column (which has an index of 3), we can type the following code:

In [20]:
sales.iloc[:, 3]

0         11/11/17
1         11/11/17
2       16/06/2017
3       18/10/2016
4       18/10/2016
           ...    
9799    28/05/2017
9800    17/01/2016
9801    17/01/2016
9802    17/01/2016
9803    17/01/2016
Name: Ship Date, Length: 9804, dtype: object

Vice versa, we can select the data corresponding to all columns of the fourth row (which has an index of 3).

In [21]:
sales.iloc[3, :]

Row ID                                                         4
Order ID                                          US-2016-108966
Order Date                                              11/10/16
Ship Date                                             18/10/2016
Shipping Method                                                3
Customer ID                                             SO-20335
Customer Name                                     Sean O'Donnell
Segment                                                 Consumer
City                                             Fort Lauderdale
State                                                    Florida
Zip Code                                                 33311.0
Region                                                     South
Product ID                                       FUR-TA-10000577
Category                                               Furniture
Sub-Category                                              Tables
Product Name       Bretfo

Using `.iloc[]` we can specifically select cells of a dataframe by inputting its row and column index values:

In [22]:
sales.iloc[234, 9]

'Florida'

Similar to slicing, we can also select specific indexes or a range of consecutive indexes:

In [23]:
sales.iloc[0:500, [3,6,12]]

Ship Date    Customer Name       Product ID
0      11/11/17      Claire Gute  FUR-BO-10001798
1      11/11/17      Claire Gute  FUR-CH-10000454
2    16/06/2017  Darrin Van Huff  OFF-LA-10000240
3    18/10/2016   Sean O'Donnell  FUR-TA-10000577
4    18/10/2016   Sean O'Donnell  OFF-ST-10000760
..          ...              ...              ...
495  31/12/2016    Maribeth Dona  OFF-EN-10001434
496  31/07/2017     Sanjit Chand  OFF-BI-10002498
497  31/07/2017     Sanjit Chand  FUR-FU-10004864
498  31/07/2017     Sanjit Chand  FUR-CH-10002602
499  31/07/2017     Sanjit Chand  FUR-FU-10000073

[500 rows x 3 columns]

**What do you think this code does? Discuss with your neighbor for approximately one minute.**

In [24]:
rows = np.arange(45,107, step = 5)

sales.iloc[rows, 13:]

Category Sub-Category  \
45   Office Supplies      Binders   
50   Office Supplies       Labels   
55   Office Supplies      Storage   
60   Office Supplies      Binders   
65         Furniture  Furnishings   
70   Office Supplies      Binders   
75   Office Supplies      Binders   
80   Office Supplies      Binders   
85         Furniture       Chairs   
90        Technology       Phones   
95   Office Supplies      Binders   
100       Technology  Accessories   
105  Office Supplies      Binders   

                                          Product Name    Sales  \
45   Wilson Jones Leather-Like Binders with DublLoc...   17.460   
50                                           Avery 485   75.180   
55                     Home/Office Personal File Carts  208.560   
60                               Plastic Binding Combs   48.480   
65                         Luxo Economy Swing Arm Lamp   79.760   
70   Avery Binding System Hidden Tab Executive Styl...    4.616   
75                                     Economy Binders    1.248   
80     Avery Heavy-Duty EZD  Binder With Locking Rings   16.740   
85                    Novimex Swivel Fabric Task Chair  301.960   
90                                  Panasonic Kx-TS550   73.584   
95   Flexible Leather- Look Classic Collection Ring...    5.682   
100  SanDisk Ultra 64 GB MicroSDHC Class 10 Memory ...   95.976   
105  Avery Trapezoid Ring Binder, 3" Capacity, Blac...   36.882   

                Full Location  
45           Eagan, Minnesota  
50        New Albany, Indiana  
55             Troy, New York  
60             Troy, New York  
65    Los Angeles, California  
70    New York City, New York  
75             Houston, Texas  
80           Decatur, Alabama  
85   Columbia, South Carolina  
90    Los Angeles, California  
95           Portland, Oregon  
100         Chicago, Illinois  
105          Aurora, Colorado

**What do you think this does? Discuss with your neighbor for approximately one minute.**

In [25]:
sales.iloc[2999:3872:100, 4::3]

Shipping Method      Segment  Zip Code         Category     Sales
2999                3     Consumer   19120.0       Technology   206.991
3099                3     Consumer   10024.0  Office Supplies    10.776
3199                3    Corporate   19134.0  Office Supplies    18.704
3299                3     Consumer   19140.0  Office Supplies    50.997
3399                3    Corporate   40475.0  Office Supplies   629.100
3499                3     Consumer   36608.0       Technology   209.970
3599                2     Consumer   19140.0       Technology   135.516
3699                2  Home Office   10701.0       Technology   163.960
3799                1    Corporate   27405.0       Technology  1359.960

Another way to extract data from dataframes is using the `.loc[]` method. The `.loc[]` method utilizes labels or boolean arrays to select data. While specifying the indices of a list of rows you wish to select, you can utilize the name of a column to extract data of interest, as shown below:

In [26]:
sales.loc[[22,485,1008], "Category"]

22      Office Supplies
485           Furniture
1008    Office Supplies
Name: Category, dtype: object

Moreover, you can select multiple columns by passing a list into the `.loc[]` method:

In [27]:
sales.loc[[22,485,1008], ["Category", "Sub-Category", "Shipping Method"]]

Category Sub-Category  Shipping Method
22    Office Supplies   Appliances                3
485         Furniture    Bookcases                1
1008  Office Supplies        Paper                3

The `.loc[]` method is a really powerful selection method that can be used to extract data based on conditions. Here, we select all rows where the `Region` column equates to `West`:

In [28]:
west = sales.loc[(sales['Region'] == 'West')]
west.head(3)

Row ID        Order ID Order Date   Ship Date  Shipping Method Customer ID  \
2       3  CA-2017-138688    12/6/17  16/06/2017                2    DV-13045   
5       6  CA-2015-115812     9/6/15  14/06/2015                3    BH-11710   
6       7  CA-2015-115812     9/6/15  14/06/2015                3    BH-11710   

     Customer Name    Segment         City       State  Zip Code Region  \
2  Darrin Van Huff  Corporate  Los Angeles  California   90036.0   West   
5  Brosina Hoffman   Consumer  Los Angeles  California   90032.0   West   
6  Brosina Hoffman   Consumer  Los Angeles  California   90032.0   West   

        Product ID         Category Sub-Category  \
2  OFF-LA-10000240  Office Supplies       Labels   
5  FUR-FU-10001487        Furniture  Furnishings   
6  OFF-AR-10002833  Office Supplies          Art   

                                        Product Name  Sales  \
2  Self-Adhesive Address Labels for Typewriters b...  14.62   
5  Eldon Expressions Wood and Plastic Desk Access...  48.86   
6                                         Newell 322   7.28   

             Full Location  
2  Los Angeles, California  
5  Los Angeles, California  
6  Los Angeles, California

We can also compound conditions using bitwise operators. We can select rows that meet one of multiple conditions by using the `|` operator:

In [29]:
w_florida = sales.loc[(sales['Region'] == 'West') | (sales['State'] == 'Florida')]
w_florida.head(3)

Row ID        Order ID Order Date   Ship Date  Shipping Method Customer ID  \
2       3  CA-2017-138688    12/6/17  16/06/2017                2    DV-13045   
3       4  US-2016-108966   11/10/16  18/10/2016                3    SO-20335   
4       5  US-2016-108966   11/10/16  18/10/2016                3    SO-20335   

     Customer Name    Segment             City       State  Zip Code Region  \
2  Darrin Van Huff  Corporate      Los Angeles  California   90036.0   West   
3   Sean O'Donnell   Consumer  Fort Lauderdale     Florida   33311.0  South   
4   Sean O'Donnell   Consumer  Fort Lauderdale     Florida   33311.0  South   

        Product ID         Category Sub-Category  \
2  OFF-LA-10000240  Office Supplies       Labels   
3  FUR-TA-10000577        Furniture       Tables   
4  OFF-ST-10000760  Office Supplies      Storage   

                                        Product Name     Sales  \
2  Self-Adhesive Address Labels for Typewriters b...   14.6200   
3      Bretford CR4500 Series Slim Rectangular Table  957.5775   
4                     Eldon Fold 'N Roll Cart System   22.3680   

              Full Location  
2   Los Angeles, California  
3  Fort Lauderdale, Florida  
4  Fort Lauderdale, Florida

If we want rows that meet <u>all</u> of our conditions of interests, we can use the `&` operator:

In [30]:
ny_low = sales.loc[(sales['State'] == 'New York') & (sales['Sales'] < 100)]
ny_low.head(3)

Row ID        Order ID  Order Date   Ship Date  Shipping Method  \
53      54  CA-2017-105816    11/12/17  17/12/2017                3   
56      57  CA-2017-111682  17/06/2017  18/06/2017                1   
58      59  CA-2017-111682  17/06/2017  18/06/2017                1   

   Customer ID    Customer Name    Segment           City     State  Zip Code  \
53    JM-15265   Janet Molinari  Corporate  New York City  New York   10024.0   
56    TB-21055  Ted Butterfield   Consumer           Troy  New York   12180.0   
58    TB-21055  Ted Butterfield   Consumer           Troy  New York   12180.0   

   Region       Product ID         Category Sub-Category  \
53   East  OFF-FA-10000304  Office Supplies    Fasteners   
56   East  OFF-PA-10001569  Office Supplies        Paper   
58   East  OFF-PA-10000587  Office Supplies        Paper   

                              Product Name  Sales            Full Location  
53                      Advantus Push Pins  15.26  New York City, New York  
56                               Xerox 232  32.40           Troy, New York  
58  Array Parchment Paper, Assorted Colors  14.56           Troy, New York

As shown before, we don't have to return all columns of the dataframe when selecting based on condition. We can return only the columns of interest by passing the names of these columns as a list into the `.loc[]` method:

In [31]:
dvh = sales.loc[(sales['Customer Name'] == 'Darrin Van Huff'), ['Order Date', 'Customer Name', 'Category']]
dvh

Order Date    Customer Name         Category
2        12/6/17  Darrin Van Huff  Office Supplies
2976     4/12/18  Darrin Van Huff        Furniture
5300  23/09/2017  Darrin Van Huff  Office Supplies
5301  23/09/2017  Darrin Van Huff  Office Supplies
5619  30/06/2018  Darrin Van Huff        Furniture
7347    11/12/18  Darrin Van Huff        Furniture
7348    11/12/18  Darrin Van Huff       Technology
7349    11/12/18  Darrin Van Huff  Office Supplies
7350    11/12/18  Darrin Van Huff        Furniture

Using a combination of index-based and label-based location, the `.loc[]` method also allows us to select specific cells of interest. By doing so, we can change data in our dataframe as needed:

In [32]:
sales.loc[0, ['Ship Date']]

Ship Date    11/11/17
Name: 0, dtype: object

In [33]:
sales.loc[0, ['Ship Date']] = '11/13/2017'
sales.head(3)

Row ID        Order ID Order Date   Ship Date  Shipping Method Customer ID  \
0       1  CA-2017-152156    8/11/17  11/13/2017                2    CG-12520   
1       2  CA-2017-152156    8/11/17    11/11/17                2    CG-12520   
2       3  CA-2017-138688    12/6/17  16/06/2017                2    DV-13045   

     Customer Name    Segment         City       State  Zip Code Region  \
0      Claire Gute   Consumer    Henderson    Kentucky   42420.0  South   
1      Claire Gute   Consumer    Henderson    Kentucky   42420.0  South   
2  Darrin Van Huff  Corporate  Los Angeles  California   90036.0   West   

        Product ID         Category Sub-Category  \
0  FUR-BO-10001798        Furniture    Bookcases   
1  FUR-CH-10000454        Furniture       Chairs   
2  OFF-LA-10000240  Office Supplies       Labels   

                                        Product Name   Sales  \
0                  Bush Somerset Collection Bookcase  261.96   
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.94   
2  Self-Adhesive Address Labels for Typewriters b...   14.62   

             Full Location  
0      Henderson, Kentucky  
1      Henderson, Kentucky  
2  Los Angeles, California

Finally, we can use `.loc[]` to insert rows in specific positions. If we want to insert a row between index 9802 and 9803, can set an index at some number in between (i.e. 9798.5) and add data using a dictionary. The key of the dictionary corresponds to the name of the column and the value corresponds to the data added in that column for the row:

In [34]:
sales.loc[9802.5] = {'Order ID': 'KB-2023-1849274', 'Order Date':'05/22/2023', "Customer Name": "Kewon Bell", 
                     "Region":"Central", "City":"Chicago", "State":"Illinois"}
sales.tail()

Row ID         Order ID  Order Date   Ship Date  Shipping Method  \
9800.0  9797.0   CA-2016-128608     12/1/16  17/01/2016              3.0   
9801.0  9798.0   CA-2016-128608     12/1/16  17/01/2016              3.0   
9802.0  9799.0   CA-2016-128608     12/1/16  17/01/2016              3.0   
9803.0  9800.0   CA-2016-128608     12/1/16  17/01/2016              3.0   
9802.5     NaN  KB-2023-1849274  05/22/2023         NaN              NaN   

       Customer ID     Customer Name    Segment     City     State  Zip Code  \
9800.0    CS-12490  Cindy Schnelling  Corporate   Toledo      Ohio   43615.0   
9801.0    CS-12490  Cindy Schnelling  Corporate   Toledo      Ohio   43615.0   
9802.0    CS-12490  Cindy Schnelling  Corporate   Toledo      Ohio   43615.0   
9803.0    CS-12490  Cindy Schnelling  Corporate   Toledo      Ohio   43615.0   
9802.5         NaN        Kewon Bell        NaN  Chicago  Illinois       NaN   

         Region       Product ID         Category Sub-Category  \
9800.0     East  OFF-AR-10001374  Office Supplies          Art   
9801.0     East  TEC-PH-10004977       Technology       Phones   
9802.0     East  TEC-PH-10000912       Technology       Phones   
9803.0     East  TEC-AC-10000487       Technology  Accessories   
9802.5  Central              NaN              NaN          NaN   

                                    Product Name    Sales Full Location  
9800.0  BIC Brite Liner Highlighters, Chisel Tip   10.368  Toledo, Ohio  
9801.0                               GE 30524EE4  235.188  Toledo, Ohio  
9802.0  Anker 24W Portable Micro USB Car Charger   26.376  Toledo, Ohio  
9803.0       SanDisk Cruzer 4 GB USB Flash Drive   10.384  Toledo, Ohio  
9802.5                                       NaN      NaN           NaN

Notice that for the columns that we did not specify, NaN was used. NaN stands for "Not a Number", which is used as a placeholder since we did not have any data for these columns. 

Now that we have our data added to the end of our dataframe, we can sort the indexes by using the `.sort_index()` method. When we call this method on our dataframe, we can set the `ascending` parameter equal to `False` to have our indexes go from largest to smallest. We can then set `inplace` equal to `True` so that it updates the dataframe without reassignment:

In [35]:
sales.sort_index(ascending=False, inplace=True)
sales.head(3)

Row ID         Order ID  Order Date   Ship Date  Shipping Method  \
9803.0  9800.0   CA-2016-128608     12/1/16  17/01/2016              3.0   
9802.5     NaN  KB-2023-1849274  05/22/2023         NaN              NaN   
9802.0  9799.0   CA-2016-128608     12/1/16  17/01/2016              3.0   

       Customer ID     Customer Name    Segment     City     State  Zip Code  \
9803.0    CS-12490  Cindy Schnelling  Corporate   Toledo      Ohio   43615.0   
9802.5         NaN        Kewon Bell        NaN  Chicago  Illinois       NaN   
9802.0    CS-12490  Cindy Schnelling  Corporate   Toledo      Ohio   43615.0   

         Region       Product ID    Category Sub-Category  \
9803.0     East  TEC-AC-10000487  Technology  Accessories   
9802.5  Central              NaN         NaN          NaN   
9802.0     East  TEC-PH-10000912  Technology       Phones   

                                    Product Name   Sales Full Location  
9803.0       SanDisk Cruzer 4 GB USB Flash Drive  10.384  Toledo, Ohio  
9802.5                                       NaN     NaN           NaN  
9802.0  Anker 24W Portable Micro USB Car Charger  26.376  Toledo, Ohio

The new row has been placed in between the 9803 index and 9802 index. This is the desired location that we wanted, however, we do not want an index as a float. To fix this, we can simply reset our indexes using `.reset_index()`. By setting `drop=True`, we get rid of the column that was once the index:

In [36]:
sales.reset_index(drop=True, inplace=True)
sales.head(3)

Row ID         Order ID  Order Date   Ship Date  Shipping Method  \
0  9800.0   CA-2016-128608     12/1/16  17/01/2016              3.0   
1     NaN  KB-2023-1849274  05/22/2023         NaN              NaN   
2  9799.0   CA-2016-128608     12/1/16  17/01/2016              3.0   

  Customer ID     Customer Name    Segment     City     State  Zip Code  \
0    CS-12490  Cindy Schnelling  Corporate   Toledo      Ohio   43615.0   
1         NaN        Kewon Bell        NaN  Chicago  Illinois       NaN   
2    CS-12490  Cindy Schnelling  Corporate   Toledo      Ohio   43615.0   

    Region       Product ID    Category Sub-Category  \
0     East  TEC-AC-10000487  Technology  Accessories   
1  Central              NaN         NaN          NaN   
2     East  TEC-PH-10000912  Technology       Phones   

                               Product Name   Sales Full Location  
0       SanDisk Cruzer 4 GB USB Flash Drive  10.384  Toledo, Ohio  
1                                       NaN     NaN           NaN  
2  Anker 24W Portable Micro USB Car Charger  26.376  Toledo, Ohio

Our dataframe has now been sorted and reset. All of the rows that were at the end are now at the beginning.

But say we made a mistake and we wanted to remove that row again, we could use the `.drop()` method again, but this time, use the `index` parameter and set it to the index of the row we want to remove. To update the dataframe you can use reassignment or set `inplace=True`:

In [37]:
sales.drop(index=1, inplace=True)
sales.head(3)

Row ID        Order ID Order Date   Ship Date  Shipping Method Customer ID  \
0  9800.0  CA-2016-128608    12/1/16  17/01/2016              3.0    CS-12490   
2  9799.0  CA-2016-128608    12/1/16  17/01/2016              3.0    CS-12490   
3  9798.0  CA-2016-128608    12/1/16  17/01/2016              3.0    CS-12490   

      Customer Name    Segment    City State  Zip Code Region  \
0  Cindy Schnelling  Corporate  Toledo  Ohio   43615.0   East   
2  Cindy Schnelling  Corporate  Toledo  Ohio   43615.0   East   
3  Cindy Schnelling  Corporate  Toledo  Ohio   43615.0   East   

        Product ID    Category Sub-Category  \
0  TEC-AC-10000487  Technology  Accessories   
2  TEC-PH-10000912  Technology       Phones   
3  TEC-PH-10004977  Technology       Phones   

                               Product Name    Sales Full Location  
0       SanDisk Cruzer 4 GB USB Flash Drive   10.384  Toledo, Ohio  
2  Anker 24W Portable Micro USB Car Charger   26.376  Toledo, Ohio  
3                               GE 30524EE4  235.188  Toledo, Ohio

So now we have the row we wanted removed, but we should still like to fix our `Row ID` column to make them integers again. We can do this using a method called `.astype()`. This method can be used to change the data type of a dataframe or series.

In [38]:
sales['Row ID'] = sales['Row ID'].astype(int)
sales.head(3)

Row ID        Order ID Order Date   Ship Date  Shipping Method Customer ID  \
0    9800  CA-2016-128608    12/1/16  17/01/2016              3.0    CS-12490   
2    9799  CA-2016-128608    12/1/16  17/01/2016              3.0    CS-12490   
3    9798  CA-2016-128608    12/1/16  17/01/2016              3.0    CS-12490   

      Customer Name    Segment    City State  Zip Code Region  \
0  Cindy Schnelling  Corporate  Toledo  Ohio   43615.0   East   
2  Cindy Schnelling  Corporate  Toledo  Ohio   43615.0   East   
3  Cindy Schnelling  Corporate  Toledo  Ohio   43615.0   East   

        Product ID    Category Sub-Category  \
0  TEC-AC-10000487  Technology  Accessories   
2  TEC-PH-10000912  Technology       Phones   
3  TEC-PH-10004977  Technology       Phones   

                               Product Name    Sales Full Location  
0       SanDisk Cruzer 4 GB USB Flash Drive   10.384  Toledo, Ohio  
2  Anker 24W Portable Micro USB Car Charger   26.376  Toledo, Ohio  
3                               GE 30524EE4  235.188  Toledo, Ohio

Finally, we can set the indexes of our dataframe to any column within the dataframe by using `.set_index()`. The `.set_index()` method can be used to set unique identifiers, such as a user ID or social security number, as this index, which could be very useful for label-based selection.

For now, we will use the `Row ID` column to set our indexes.

In [39]:
sales.set_index('Row ID', inplace=True)
sales.tail(3)

Order ID Order Date   Ship Date  Shipping Method Customer ID  \
Row ID                                                                       
3       CA-2017-138688    12/6/17  16/06/2017              2.0    DV-13045   
2       CA-2017-152156    8/11/17    11/11/17              2.0    CG-12520   
1       CA-2017-152156    8/11/17  11/13/2017              2.0    CG-12520   

          Customer Name    Segment         City       State  Zip Code Region  \
Row ID                                                                         
3       Darrin Van Huff  Corporate  Los Angeles  California   90036.0   West   
2           Claire Gute   Consumer    Henderson    Kentucky   42420.0  South   
1           Claire Gute   Consumer    Henderson    Kentucky   42420.0  South   

             Product ID         Category Sub-Category  \
Row ID                                                  
3       OFF-LA-10000240  Office Supplies       Labels   
2       FUR-CH-10000454        Furniture       Chairs   
1       FUR-BO-10001798        Furniture    Bookcases   

                                             Product Name   Sales  \
Row ID                                                              
3       Self-Adhesive Address Labels for Typewriters b...   14.62   
2       Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.94   
1                       Bush Somerset Collection Bookcase  261.96   

                  Full Location  
Row ID                           
3       Los Angeles, California  
2           Henderson, Kentucky  
1           Henderson, Kentucky

### <code style="background:#83ebd5;color:black">Exercise:</code>

- The `Zip Code` column of `sales`, for some odd reason, is recorded as floats. Convert this column to integers.

- From `sales`, select all entries from Southwestern states (Arizona, New Mexico, Texas, Oklahoma) that ordered products from the `Art` sub-category. Assign this dataframe as `southwest`. *Hint: this may require an intermediate step.*

- From `sales`, select the `Customer ID`, `Customer Name`, `Segment`, and `Sales` columns. Assign this dataframe as `customers`. Then, set the index of `customers` to be the `Customer ID` column. Using the index, select orders with the customer ID <u>SH-19975</u>. *Hint: refer to the `pandas.DataFrame.loc` online documentation if you get stuck.*

## Organizing, manipulating, analyzing data in dataframes

There are many functions and methods that can assist with organizing, manipulating, and analyzing data in dataframes. We will discuss several important ones to remember.

The first method we will discuss is the `.sort_values()` method. This is a useful method for looking at trends or sequential values in a dataset. The `.sort_values()` method has an `ascending` parameter, which is set to `True` by default. It also has an `inplace` parameter, which is set to `False` by default: 



In [40]:
sales.sort_values('City', inplace= True)
sales.head(3)

Order ID  Order Date   Ship Date  Shipping Method Customer ID  \
Row ID                                                                        
9261    CA-2018-167976    11/11/18  14/11/2018              2.0    JL-15505   
6990    CA-2018-165099    11/12/18  13/12/2018              1.0    DK-13375   
454     CA-2018-147277  20/10/2018  24/10/2018              3.0    EB-13705   

          Customer Name    Segment      City         State  Zip Code   Region  \
Row ID                                                                          
9261    Jeremy Lonsdale   Consumer  Aberdeen  South Dakota   57401.0  Central   
6990        Dennis Kane   Consumer   Abilene         Texas   79605.0  Central   
454          Ed Braxton  Corporate     Akron          Ohio   44312.0     East   

             Product ID         Category Sub-Category  \
Row ID                                                  
9261    OFF-SU-10004661  Office Supplies     Supplies   
6990    OFF-AP-10001634  Office Supplies   Appliances   
454     FUR-TA-10001539        Furniture       Tables   

                                        Product Name    Sales  \
Row ID                                                          
9261                   Acme Titanium Bonded Scissors   25.500   
6990    Hoover Commercial Lightweight Upright Vacuum    1.392   
454         Chromcraft Rectangular Conference Tables  284.364   

                 Full Location  
Row ID                          
9261    Aberdeen, South Dakota  
6990            Abilene, Texas  
454                Akron, Ohio

Another useful method is the 
`.value_counts()` method. This method can be used on a series to count the number of occurences for each unique value in the series. Let's use this method to determine how many shipments were shipped to each region:






In [41]:
sales['Region'].value_counts()

Region
West       3141
East       2788
Central    2277
South      1598
Name: count, dtype: int64

There are a number of statistical and summative methods that can be used on dataframes as well. For example, the `.min()` and `.max()` methods can be used on a a dataframe or a column within a dataframe to determine the minimum and maximum values, respectively. The `.describe()` method returns multiple useful summary statistics:

In [42]:
sales['Sales'].min()

0.444

In [43]:
sales['Sales'].max()

22638.48

In [ ]:
sales['Sales'].describe()

Here's a short list of methods that can be used to determine summative information and/or statistics:

`.min()` - determines the minimum value in a dataframe/series

`.max()` - determines the maximum value in a dataframe/series

`.mean()` - determines the mean value of a dataframe/series

`.median()` - determines the median value in a dataframe/series

`.sum()` - determines the sum of the values in a dataframe/series

`.mode()` - determines the mode of a dataframe/series

`.nlargest()` - determines the largest *n* items in a series. (Default = 5)

`.nsmallest()` - determines the smallest *n* items in a series. (Default = 5)

`.count()` - counts all non-null items in a dataframe/series

`.value_counts` - counts the number of times an item occurs in a dataframe/series

`.nunique()` - counts the number of unique items in a dataframe/series

`.unique()` - lists all of the unique items in a dataframe/series

`.describe` - generates descriptive statistics of a dataframe column, including those that summarize the central tendency, dispersion and shape of a dataset’s distribution, excluding null values (NaN).

Sometimes when working with data, especially data generated by human input, we encounter errors within the data. One useful tool that can be used to find mistakes is the `.duplicated()` method. This method shows duplicated rows within a dataframe, like so:

In [ ]:
sales[sales.duplicated()]

When `.duplicated()` is called on a dataframe, another dataframe containing the rows of the duplicated entries is returned to us. We can easily get rid of these duplicated rows by calling `.drop_duplicates()` on the dataframe:

In [ ]:
print(sales.shape)
sales = sales.drop_duplicates()
print(sales.shape)

We see that prior to dropping the duplicates, `sales` had 9804 rows. After dropping the duplicates, there are now 9799 rows, 5 less than before. This is the number of rows that were determined to be duplicated in `sales`, as shown above.

We can also correct typos observed in dataframes. Say for example, we had a customer listed in our dataframe as Eric Hoffmann. We can extract all the rows pertaining to this customer:

In [ ]:
sales.loc[(sales['Customer Name']) == 'Eric Hoffmann'].head(3)

Say an extra 'n' was accidentally added to Eric's last name. We can correct this typo in all of these cells by using the `.replace()` method. When using the `.replace()` method, we can call the current mispelling into the `to_replace` parameter. We can then call the correct spelling into the `value` parameter. The `inplace` parameter will update our dataframe in real time. 

If we try to extract rows with the original spelling, we get an empty dataframe returned to us:

In [ ]:
sales["Customer Name"].replace(to_replace='Eric Hoffmann', value='Eric Hoffman', inplace=True)

sales.loc[(sales['Customer Name']) == 'Eric Hoffmann'].head(3)

If we use the new spelling, we can see that the name has been updated:

In [ ]:
sales.loc[(sales['Customer Name']) == 'Eric Hoffman'].head(3)

We can also apply string methods to columns in a dataframe and update the column through variable assignment:

In [ ]:
sales['State'] = sales['State'].str.upper()
sales.head(3)

Another useful method is the `.map()` method. The `.map()` method substitutes values from a series, dictionary, or function based on current values in the dataframe. Using this method, we can map the values in the `Shipping Method` column to their corresponding string values like so:

In [ ]:
new_vals = {0: 'Same Day', 1: 'First Class', 2:'Second Class', 3:'Standard Class'}

sales['Shipping Method'] = sales['Shipping Method'].map(new_vals)
sales.head()

Applying values and functions en masse to datasets is a quick and easy way to analyze and modify data. The `.applymap()` method is great for this because uses a function and applies it to an entire dataframe. Say we wanted to convert all the data in sales to string data, we would do:

In [ ]:
strings = sales.applymap(str)
strings.head(3)

By eye, `strings` looks the same as `sales`, but upon closer investigation, we can see that the numbers have been converted to strings:

In [ ]:
strings['Zip Code'][5] #Picking a random number from the Zip Code column

In [ ]:
print(type(strings['Zip Code'][5]))

The `.apply()` method is also great for applying a function to a series or multiple columns within a dataframe across an axis (i.e. across rows or across columns). For example, if we wanted to count the characters in each state for each row, we could apply the `len` function:

In [ ]:
sales['State'].apply(len)

When using the`.apply()` method on multiple columns of a dataframe, the`axis` parameter dictates if you function is applied acrossed rows or columns. When `axis=0`, the function is applied across columns, and when `axis=1`, the function is applied across rows.

<img src="https://raw.githubusercontent.com/campbelle1/CAN2023/4c78054e6a24add10344d95ab609c1ecba3d6999/apply-method.png" width="600"/>

Apply can be useful for aggregate functions, such as `np.min`, `np.max`, `np.mean`, and `np.sum`. For more information, refer to the `.apply()` method documentaion in the `pandas`. You will see other applications of `.apply()` when learning how to make user-defined functions.

# Activity

A study including 1040 subjects examines the effects of a weight loss medication. The study records subject demographics, such as age and sex, as well as each subject's weight for seven days. Each subject was assigned a unique identification number.

- Load the data from the provided internet URL. Assign this data to a dataframe called `study`.

- The downloaded data file did not include data from the last two subjects. Using the following information, append this data to `study`:
<img align="center" src="https://raw.githubusercontent.com/campbelle1/CAN2023/161d2a06dfe210825ea764fb65d0c5cfdcdd51a7/Extras.png" width="20%"/> 

- From `study`, determine the minimum and maxmimum weights for each subject within the seven day period of the experiment. Append this information as new columns called `Min Weight` and `Max Weight` in the `study` dataframe.

- Find the net weight change from the beginning of the experiment to the end of the experiment for each subject.

- For the ease of data collection, treatments for the study were recorded as 0 (to mean the subject received the placebo) or 1 (to mean the subject received the weight loss medication). Replace these integers with the description of the treatment (that is, *Placebo* or *Medication*).

- Researchers have seen anecdotally that older subjects seem to lose more weight than younger subjects. Determine the top 10% of subjects that had the most weight loss. Calculate the proportion of subjects in the top 10% that were over the age of 35.